README.txt
Introduction
============

This product enhances sc.social.like with a portlet and others social plugins.


docs/HISTORY.txt
Changelog
=========

1.3 (unreleased)
----------------

- added a viewlet with the js [sithmel]
- added a portlet class (vertical or horizontal) [sithmel]
- fix vertical/horizontal on pinterest and linkedin [sithmel]


1.2 (2012-04-03)
----------------

- Fixed readme and history


1.0 (2012-01-12)
----------------

- added linkedin
- overridden viewlet 



redomino/social/tests/base.py
from Products.Five import zcml
from Products.Five import fiveconfigure

from Testing import ZopeTestCase as ztc

from Products.PloneTestCase import PloneTestCase as ptc
from Products.PloneTestCase.layer import onsetup


@onsetup
def setup_product():
    """Set up additional products and ZCML required to test this product.

    The @onsetup decorator causes the execution of this body to be deferred
    until the setup of the Plone site testing layer.
    """

    # Load the ZCML configuration for this package and its dependencies

    fiveconfigure.debug_mode = True
    import sc.social.like
    import redomino.social
    zcml.load_config('configure.zcml', sc.social.like)
    zcml.load_config('configure.zcml', redomino.social)
    fiveconfigure.debug_mode = False

    # We need to tell the testing framework that these products
    # should be available. This can't happen until after we have loaded
    # the ZCML.

    ztc.installPackage('redomino.social')

# The order here is important: We first call the deferred function and then
# let PloneTestCase install it during Plone site setup

setup_product()
ptc.setupPloneSite(products=['redomino.social'])


class TestCase(ptc.PloneTestCase):
    """Base class used for test cases
    """


class FunctionalTestCase(ptc.FunctionalTestCase):
    """Test case class used for functional (doc-)tests
    """


redomino/social/tests/test_portlet.py
# Authors: Andrea D'Este <andrea.deste@redomino.com>
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.

from zope.component import getUtility, getMultiAdapter

from plone.portlets.interfaces import IPortletType
from plone.portlets.interfaces import IPortletManager
from plone.portlets.interfaces import IPortletAssignment
from plone.portlets.interfaces import IPortletDataProvider
from plone.portlets.interfaces import IPortletRenderer

from plone.app.portlets.storage import PortletAssignmentMapping

from redomino.social.portlets import socialportlet

from redomino.social.tests.base import TestCase
import unittest2 as unittest


class TestPortlet(TestCase):

    def afterSetUp(self):
        self.setRoles(('Manager', ))

    def test_portlet_type_registered(self):
        portlet = getUtility(
            IPortletType,
            name='redomino.social.portlets.SocialPortlet')
        self.assertEquals(portlet.addview,
                          'redomino.social.portlets.SocialPortlet')

    def test_interfaces(self):
        portlet = socialportlet.Assignment()
        self.failUnless(IPortletAssignment.providedBy(portlet))
        self.failUnless(IPortletDataProvider.providedBy(portlet.data))

    def test_invoke_add_view(self):
        portlet = getUtility(
            IPortletType,
            name='redomino.social.portlets.SocialPortlet')
        mapping = self.portal.restrictedTraverse(
            '++contextportlets++plone.leftcolumn')
        for m in mapping.keys():
            del mapping[m]
        addview = mapping.restrictedTraverse('+/' + portlet.addview)

        addview()

        self.assertEquals(len(mapping), 1)
        self.failUnless(isinstance(mapping.values()[0],
                                   socialportlet.Assignment))

    def test_obtain_renderer(self):
        context = self.folder
        request = self.folder.REQUEST
        view = self.folder.restrictedTraverse('@@plone')
        manager = getUtility(IPortletManager, name='plone.rightcolumn',
                             context=self.portal)

        assignment = socialportlet.Assignment()

        renderer = getMultiAdapter(
            (context, request, view, manager, assignment), IPortletRenderer)
        self.failUnless(isinstance(renderer, socialportlet.Renderer))


class TestRenderer(TestCase):

    def afterSetUp(self):
        self.setRoles(('Manager', ))

    def renderer(self, context=None, request=None, view=None, manager=None,
                 assignment=None):
        context = context or self.folder
        request = request or self.folder.REQUEST
        view = view or self.folder.restrictedTraverse('@@plone')
        manager = manager or getUtility(
            IPortletManager, name='plone.rightcolumn', context=self.portal)

        assignment = assignment or socialportlet.Assignment()
        return getMultiAdapter((context, request, view, manager, assignment),
                               IPortletRenderer)

    def test_render(self):
        r = self.renderer(context=self.portal,
                          assignment=socialportlet.Assignment())
        r = r.__of__(self.folder)
        r.update()
        output = r.render()
        self.failUnless('portlet-social-like' in output)


def test_suite():
    from unittest import TestSuite, makeSuite
    suite = TestSuite()
    suite.addTest(makeSuite(TestPortlet))
    suite.addTest(makeSuite(TestRenderer))
    return suite


redomino/social/__init__.py
# Authors: Andrea D'Este <andrea.deste@redomino.com>
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.

from zope.i18nmessageid import MessageFactory
                                                                         
_ = MessageFactory('redomino.social')

def initialize(context):
    """Initializer called when used as a Zope 2 product."""


redomino/social/controlpanel/likes.py
# Authors: Davide Moro <davide.moro@redomino.com>
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.

from sc.social.like.controlpanel.likes import BaseControlPanelAdapter
from sc.social.like.controlpanel.likes import ProvidersControlPanel as OriginalProvidersControlPanel
from sc.social.like.controlpanel.likes import IProvidersSchema, ITwitterSchema, IFbSchema, IGpSchema, MultiSelectWidget
from sc.social.like.controlpanel.likes import baseset, twitterset, fbset, gpset

from zope.interface import Interface
from zope.schema import Bool
from Products.CMFPlone.interfaces import IPloneSiteRoot
from Products.CMFDefault.formlib.schema import ProxyFieldProperty
from plone.fieldsets.fieldsets import FormFieldsets
from plone.app.controlpanel.form import ControlPanelForm
from sc.social.like import LikeMessageFactory as _
from zope.component import adapts
from zope.interface import implements

class ILinkedinSchema(Interface):
    """ Linkedin configurations """

    linkedin_enabled = Bool(
        title=_(u"Enable Linkedin button"),
        default=True,
        required=False,
        )

class LinkedinControlPanelAdapter(BaseControlPanelAdapter):
    """ Linkedin control panel adapter """
    adapts(IPloneSiteRoot)
    implements(ILinkedinSchema)

    linkedin_enabled = ProxyFieldProperty(ILinkedinSchema['linkedin_enabled'])

linkedinset = FormFieldsets(ILinkedinSchema)
linkedinset.id = 'linkedinset'
linkedinset.label = _(u'Linkedin settings')

class IPinterestSchema(Interface):
    """ Pinterest configurations """

    pinterest_enabled = Bool(
        title=_(u"Enable Pinterest button"),
        default=True,
        required=False,
        )

class PinterestControlPanelAdapter(BaseControlPanelAdapter):
    """ Pinterest control panel adapter """
    adapts(IPloneSiteRoot)
    implements(IPinterestSchema)

    pinterest_enabled = ProxyFieldProperty(IPinterestSchema['pinterest_enabled'])

pinterestset = FormFieldsets(IPinterestSchema)
pinterestset.id = 'pinterestset'
pinterestset.label = _(u'Pinterest settings')

class ProvidersControlPanel(OriginalProvidersControlPanel):
    """ """
    form_fields = FormFieldsets(baseset, twitterset, fbset, gpset, linkedinset, pinterestset)

    form_fields['enabled_portal_types'].custom_widget = MultiSelectWidget




redomino/social/portlets/socialportlet.py
# Authors: Andrea D'Este <andrea.deste@redomino.com>
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.

from zope.interface import Interface
from zope.interface import implements       
                                                    
from plone.app.portlets.portlets import base
from plone.portlets.interfaces import IPortletDataProvider                    
from Products.CMFCore.utils import getToolByName  
from zope.component import getMultiAdapter
from zope import schema                    
from zope.formlib import form            
from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile
                 
from redomino.social import _
                                                    
from zope.i18nmessageid import MessageFactory  
__ = MessageFactory("plone")       
                                         
class ISocialPortlet(IPortletDataProvider):
    """A portlet                         
                                           
    It inherits from IPortletDataProvider because for this portlet, the
    data that is being rendered and the portlet assignment itself are the     
    same.    
    """                                                                       
                                                                              
#    portlet_title = schema.TextLine(title=__(u'Title'),                       
#                                  description=_(u'...'),                      
#                                  required=True,                              
#                                  default=u'')                                
                                                                              
class Assignment(base.Assignment):                                            
    """Portlet assignment.                                                    
                                                                              
    This is what is actually managed through the portlets UI and associated   
    with columns.                                                             
    """                                                                       
                                                                              
    implements(ISocialPortlet)                                              
#    portlet_title = u""                                                       
#                                                                              
#                                                                              
#    def __init__(self, portlet_title=u''):
#        self.portlet_title = portlet_title

    @property
    def title(self):
        """This property is used to give the title of the portlet in the
        "manage portlets" screen.
        """
        return _(u"Social Portlet")


class Renderer(base.Renderer):
    """Portlet renderer """
    render = ViewPageTemplateFile('socialportlet.pt')

    def __init__(self, context, request, view, manager, data):
        base.Renderer.__init__(self, context, request, view, manager, data)
        self.membership = getToolByName(self.context, 'portal_membership')
        self.portal_state = getMultiAdapter((context, request), name=u'plone_portal_state')
        urltool = getToolByName(context, "portal_url")
        self.portal  = urltool.getPortalObject()
        
        pp = getToolByName(context, 'portal_properties')
        self.sheet = getattr(pp, 'sc_social_likes_properties', None)


        self.enabled_portal_types = self.sheet.getProperty("enabled_portal_types", [])
        self.typebutton = self.sheet.getProperty("typebutton", "")
        self.twitter_enabled = self.sheet.getProperty("twitter_enabled", True)
        self.twittvia = self.sheet.getProperty("twittvia", "")
        self.fb_enabled = self.sheet.getProperty("fb_enabled", True)
        self.fbaction = self.sheet.getProperty("fbaction", "")
        self.fbadmins = self.sheet.getProperty("fbadmins", "")
        self.gp_enabled = self.sheet.getProperty("gp_enabled", True)
        self.linkedin_enabled = self.sheet.getProperty("linkedin_enabled", True)
        self.pinterest_enabled = self.sheet.getProperty("pinterest_enabled", True)

    @property
    def title(self):
        """return title of feed for portlet"""
        return getattr(self.data, 'portlet_title', '')

    def isAnon(self):
        """ return True if member is Anonymous"""
        if not self.portal_state.anonymous():
            return False
        return True

    @property
    def available(self):
        context = self.context
        enabled_portal_types = self.enabled_portal_types
        return context.portal_type in enabled_portal_types

    def classes(self):
        return "portlet portletSocialPortlet %s" % self.typebutton

class AddForm(base.NullAddForm):
    """Portlet add form. """

    def create(self):
        return Assignment()
